File: python/프로그래머스/게임맵최단거리.py
from collections import deque

def solution(maps):
    n = len(maps[0])
    m = len(maps)
    
    # 도착점이 벽일 경우 바로 -1 반환
    if maps[m - 1][n - 1] == 0:
        return -1

    # BFS 초기화
    queue = deque([(0, 0, 1)])  # (x, y, depth)
    visited = [[False for _ in range(n)] for _ in range(m)]
    visited[0][0] = True
    
    moves = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    while queue:
        x, y, depth = queue.popleft()

        # 도착점에 도달했을 때
        if x == n - 1 and y == m - 1:
            return depth

        # 네 방향으로 이동
        for move in moves:
            px, py = x + move[0], y + move[1]

            if 0 <= px < n and 0 <= py < m and not visited[py][px] and maps[py][px] == 1:
                visited[py][px] = True  # 방문 기록
                queue.append((px, py, depth + 1))

    # 도착점에 도달하지 못한 경우
    return -1

File: java/게임맵최단거리.java
import java.util.*;

class Solution {
    int min = Integer.MAX_VALUE;
    int[] target = new int[2];
    int[][] moves = new int[][]{{1,0},{0,1},{-1,0},{0,-1}};
    boolean[][] visited;
    int[][] myMaps;
    public int solution(int[][] maps) {
        int answer = 0;
        target[0] = maps.length;
        target[1] = maps[0].length;
        visited = new boolean[maps.length][maps[0].length];
        myMaps = maps;
        bfs();
        if (min == Integer.MAX_VALUE) min = -1;
        return min;
    } 
    
//     private void dfs(int x, int y, int count) {
//         if (min <= count) {
//             return;
//         }
//         visited[x][y] = true;
        
//         if (x == target[0]-1 && y == target[1]-1) {
//             min = Math.min(min, count);
//             return;
//         }
        
//         for(int[]mv: moves) {
//             int nx = x + mv[0];
//             int ny = y + mv[1];
            
//             if (valid(nx,ny)) {
//                 dfs(nx,ny,count+1);
//                 visited[nx][ny] = false;
//             }
//         }
//     }    
    private void bfs() {
        Deque<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0,0,1});
        visited[0][0] = true;
        
        while(!q.isEmpty()) {
            int[] current = q.poll();
            int x = current[0];
            int y = current[1];
            int count = current[2];
        
            if (x==target[0]-1&&y==target[1]-1) {
                min = count;
                break;
            }
            
            for(int[]mv: moves) {
                int nx = x + mv[0];
                int ny = y + mv[1];

                if (valid(nx,ny)) {
                    q.offer(new int[]{nx,ny,count+1});
                    visited[nx][ny] = true;
                }
            }
        }
    }  
    
    private boolean valid(int x,int y) {
        if (0<= x && x < visited.length && 0 <= y && y < visited[0].length) {
            if(!visited[x][y] && myMaps[x][y] == 1) return true;
        }
        return false;
    }
}
